web/swift/swift.go

```diff
@@ -0,0 +1,239 @@
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+
+	"github.com/cozy/cozy-stack/model/instance"
+	"github.com/cozy/cozy-stack/model/instance/lifecycle"
+	"github.com/cozy/cozy-stack/pkg/config/config"
+	"github.com/labstack/echo/v4"
+	"github.com/ncw/swift"
+)
+
+func ListLayouts(c echo.Context) error {
+	type layout struct {
+		Counter int      `json:"counter"`
+		Domains []string `json:"domains,omitempty"`
+	}
+	var layoutV1, layoutV2a, layoutV2b, layoutUnknown, layoutV3 layout
+
+	flagShowDomains := false
+	flagParam := c.QueryParam("show_domains")
+	if converted, err := strconv.ParseBool(flagParam); err == nil {
+		flagShowDomains = converted
+	}
+
+	instances, err := instance.List()
+	if err != nil {
+		return err
+	}
+	for _, inst := range instances {
+		switch inst.SwiftLayout {
+		case 0:
+			layoutV1.Counter++
+			if flagShowDomains {
+				layoutV1.Domains = append(layoutV1.Domains, inst.Domain)
+			}
+		case 1:
+			switch inst.DBPrefix() {
+			case inst.Domain:
+				layoutV2a.Counter++
+				if flagShowDomains {
+					layoutV2a.Domains = append(layoutV2a.Domains, inst.Domain)
+				}
+			case inst.Prefix:
+				layoutV2b.Counter++
+				if flagShowDomains {
+					layoutV2b.Domains = append(layoutV2b.Domains, inst.Domain)
+				}
+			default:
+				layoutUnknown.Counter++
+				if flagShowDomains {
+					layoutUnknown.Domains = append(layoutUnknown.Domains, inst.Domain)
+				}
+			}
+		case 2:
+			layoutV3.Counter++
+			if flagShowDomains {
+				layoutV3.Domains = append(layoutV3.Domains, inst.Domain)
+			}
+		default:
+			layoutUnknown.Counter++
+			if flagShowDomains {
+				layoutUnknown.Domains = append(layoutUnknown.Domains, inst.Domain)
+			}
+		}
+	}
+
+	output := make(map[string]interface{})
+	output["v1"] = layoutV1
+	output["v2a"] = layoutV2a
+	output["v2b"] = layoutV2b
+	output["unknown"] = layoutUnknown
+	output["v3"] = layoutV3
+	output["total"] = layoutV1.Counter + layoutV2a.Counter + layoutV2b.Counter + layoutUnknown.Counter + layoutV3.Counter
+
+	return c.JSON(http.StatusOK, output)
+}
+
+// GetObject retrieves a Swift object from an instance
+func GetObject(c echo.Context) error {
+	domain := c.Param("domain")
+	object := c.Param("object")
+	unescaped, err := url.PathUnescape(object)
+	if err != nil {
+		return err
+	}
+
+	i, err := lifecycle.GetInstance(domain)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("test", unescaped)
+	buf := new(bytes.Buffer)
+	sc := config.GetSwiftConnection()
+	_, err = sc.ObjectGet(swiftContainer(i), unescaped, buf, false, nil)
+	if err != nil {
+		return err
+	}
+
+	out := struct {
+		Content string `json:"content"`
+	}{
+		buf.String(),
+	}
+
+	return c.JSON(http.StatusOK, out)
+}
+
+// PutObject puts an object into Swift
+func PutObject(c echo.Context) error {
+	type reqStruct struct {
+		Instance    string `json:"instance"`
+		ObjectName  string `json:"object_name"`
+		Content     string `json:"content"`
+		ContentType string `json:"content_type"`
+	}
```

Comment:
Idem, I'd prefer not to use JSON here. The body of the request should be the content of the file, and nothing else. For the content-type, an HTTP header is the right way to send it. For the instance, it can be the `Host` HTTP header. For the ObjectName, it can be in the path or in the query string.